logstash-core/src/main/java/org/logstash/config/ir/ConfigCompiler.java

```diff
@@ -42,17 +50,75 @@ private ConfigCompiler() {
+        Map<PluginDefinition.Type, List<Graph>> groupedPipelineSections = sourcesWithMetadata.stream()
+                .map(swm -> compileGraph(swm, supportEscapes))
+                .flatMap(m -> m.entrySet().stream())
+                .filter(e -> e.getValue() != null)
+                .collect(groupingBy(Map.Entry::getKey,
+                            mapping(Map.Entry::getValue, toList())));
+
+        Graph inputGraph = Graph.combine(groupedPipelineSections.get(PluginDefinition.Type.INPUT).toArray(new Graph[0])).graph;
+        Graph outputGraph = Graph.combine(groupedPipelineSections.get(PluginDefinition.Type.OUTPUT).toArray(new Graph[0])).graph;
+        Graph filterGraph = groupedPipelineSections.get(PluginDefinition.Type.FILTER).stream()
+                .reduce(ConfigCompiler::chainWithUntypedException).orElse(null);
+
+        String originalSource = sourcesWithMetadata.stream().map(SourceWithMetadata::getText).collect(Collectors.joining("\n"));
+        return new PipelineIR(inputGraph, filterGraph, outputGraph, originalSource);
+    }
+
+    private static Graph chainWithUntypedException(Graph g1, Graph g2) {
+        try {
+            return g1.chain(g2);
+        } catch (InvalidIRException iirex) {
+            throw new IllegalArgumentException(iirex);
+        }
+    }
+
+    private static Map<PluginDefinition.Type, Statement> compileImperative(SourceWithMetadata sourceWithMetadata,
+                                                                           boolean supportEscapes) {
         final IRubyObject compiler = RubyUtil.RUBY.executeScript(
                 "require 'logstash/compiler'\nLogStash::Compiler",
                 ""
         );
-        final IRubyObject code =
-            compiler.callMethod(RubyUtil.RUBY.getCurrentContext(), "compile_sources",
-                new IRubyObject[]{sourcesWithMetadata, RubyUtil.RUBY.newBoolean(supportEscapes)}
-            );
-        return code.toJava(PipelineIR.class);
+        // invoke Ruby interpreter to execute LSCL treetop
+        final IRubyObject code = compiler.callMethod(RubyUtil.RUBY.getCurrentContext(), "compile_imperative",
+                new IRubyObject[]{
+                        JavaUtil.convertJavaToRuby(RubyUtil.RUBY, sourceWithMetadata),
+                        RubyUtil.RUBY.newBoolean(supportEscapes)
+                });
+        RubyHash hash = (RubyHash) code;
+        Map<PluginDefinition.Type, Statement> result = new HashMap<>();
+        result.put(PluginDefinition.Type.INPUT, readStatementFromRubyHash(hash, "input"));
+        result.put(PluginDefinition.Type.FILTER, readStatementFromRubyHash(hash, "filter"));
+        result.put(PluginDefinition.Type.OUTPUT, readStatementFromRubyHash(hash, "output"));
+        return result;
+    }
+
+    private static Statement readStatementFromRubyHash(RubyHash hash, String key) {
+        IRubyObject inputValue = hash.op_aref(RubyUtil.RUBY.getCurrentContext(), RubyUtil.RUBY.newSymbol(key));
```

Comment:
here I think using:
```suggestion
        IRubyObject inputValue = hash.fastARef(RubyUtil.RUBY.newSymbol(key));
```
makes sense as we assume the value always to be found, so a `NPE` (method returns `null` instead of `nil`) 
would indicate a bug instead of silently `nil.toJava` converting to `null` ...